Add manifest extension AoT

JacobSzwejbka opened the pull request:
Summary:
Add some infra for us to optionally add some key structured data to the end of a pte. This diff is around enabling users to easily tag their model with a cryptographic signature. Has room to expand later.

A key design motivation is it would be ideal if this is transparent to the rest of the extensions we have today. Im claiming the prime footer real estate for this which is unused today by anything in tree. This should let it be composable with other formats like bundledProgram too.

Differential Revision: D82052721

lucylq commented on extension/manifest/_manifest.py:
```diff
@@ -0,0 +1,154 @@
+
+ def to_bytes(self) -> bytes:
+     """Returns the binary representation of the Manifest. Written
+     bottom up.
```
nit, explain why it's written bottom up

"Returns the binary representation of the Manifest. Written bottom up to allow for BC considerations. The compatibility-preserving way to make changes is to increase the header's length field and add new fields at the top. This means we can always check the last n bytes for the magic and size, and then load the full footer."

lucylq commented on extension/manifest/_manifest.py:
```diff
@@ -0,0 +1,154 @@
+         +self.padding_size.to_bytes(4, byteorder=_MANIFEST_BYTEORDER)
+         # uint32_t: Size of this manifest. This makes it easier to add new
+         # fields to this header in the future. Always use the proper size
+         # (i.e., ignore self.length) since there's no reason to create an
```
What is self.length used for?

lucylq commented on extension/manifest/_manifest.py:
```diff
@@ -0,0 +1,154 @@
+ padding_size: int = 0
+
+ # Size in bytes between the top of the manifest and the start of the data it was appended to.
+ program_offset: int = 0
```
Maybe 'pte_offset'? program could refer to pte without segments.

lucylq commented on extension/manifest/_manifest.py:
```diff
@@ -0,0 +1,154 @@
+
+ signature: int
+
+ # Size in bytes between the top of the manifest and the end of the data it was appended to
```
nit: add "The PTE data is aligned to 16 by default."

GregoryComer commented:
Is there a strong reason to not include this in the core AOT code (as opposed to a dedicated extension?). I don't have a super strong opinion on this, but I do worry about the growing number of fine-grained extensions being confusing to users and compromising UX.

JacobSzwejbka committed:
Add manifest extension AoT (#14128)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/14128

Add some infra for us to optionally add some key structured data to the end of a pte. This diff is around enabling users to easily tag their model with a cryptographic signature. Has room to expand later.

A key design motivation is it would be ideal if this is transparent to the rest of the extensions we have today. Im claiming the prime footer real estate for this which is unused today by anything in tree. This should let it be composable with other formats like bundledProgram too.

Differential Revision: D82052721

mergennachin commented on extension/manifest/_manifest.py:
```diff
@@ -0,0 +1,206 @@
+# regardless of the host system, since all commonly-used modern CPUs are little
+# endian.
+_MANIFEST_BYTEORDER: Literal["little"] = "little"
+
+
```
Write a docblock about an example usage of manifest file for higher layer consumers. Also mention that manifest is a mechanism, not a security policy.  And explicitly say that consumers implements appropriate security for their threat model

```
# 1. Generate PTE file
pte_data = serialize_pte_binary(program)

# 2. Create cryptographic signature of PTE data
signature = sign_with_private_key(pte_data, private_key)  # e.g., RSA, ECDSA

# 3. Append manifest with signature
manifest = Manifest(signature=signature)
pte_with_manifest = append_manifest(pte_data, manifest)

Verification Process

# 1. Extract manifest from end of file
manifest = Manifest.from_bytes(file_data)

# 2. Extract PTE data (using program_offset)
pte_data = file_data[:-(manifest_length + padding)]

# 3. Verify signature with public key
is_valid = verify_signature(pte_data, manifest.signature, public_key)

```

mergennachin commented on extension/manifest/_manifest.py:
```diff
@@ -0,0 +1,206 @@
+    # Unique ID for the data the manifest was appended to. Often this might contain
+    # a crytographic signature for the data.
+    signature: bytes
+
```
Add version

mergennachin replied:
I can imagine people can use this for other use-cases besides security, such as saving arbitrary serializable metadata.

For instance, saving tokenizer.json file location etc.

mergennachin commented on extension/manifest/_manifest.py:
```diff
@@ -0,0 +1,206 @@
+    # Unique ID for the data the manifest was appended to. Often this might contain
+    # a crytographic signature for the data.
+    signature: bytes
+
```
Should you add timestamp field too?

mergennachin replied:
Something like this?

```
  @dataclass
  class Manifest:
      type: str  # "signature", "checksum", "metadata", etc.
      version: int
      payload: bytes 
      timestamp: Optional[int]
      attributes: Dict[str, str] 
```

mergennachin commented on extension/manifest/_manifest.py:
```diff
@@ -0,0 +1,206 @@
+
+    EXPECTED_MAGIC: ClassVar[bytes] = b"em00"
+
+    MAX_SIGNATURE_SIZE: ClassVar[int] = 512
```
why is this fixed?

mergennachin commented on extension/manifest/_manifest.py:
```diff
@@ -0,0 +1,206 @@
+        return data
+
+    @staticmethod
+    def from_bytes(data: bytes) -> "_ManifestLayout":
```
For large files you have to read the whole thing?

JacobSzwejbka replied:
Could add more methods like from file that just load the last MAX_SIZE bytes. Or stream.

JacobSzwejbka replied:
I dont really expect people to be verifying the hash in python though. Its mostly just there for testing.
